File: packages/data_ts/src/repository_factory.ts
import type { DemoCatalog } from "./demo_repository";
import { DemoRepository } from "./demo_repository";
import { ProductionRepository } from "./production_repository";
import type { AssalDataSourceMode, AssalRepository, ProductionSelectGateway } from "./repository";

export function createAssalRepository(options: {
  mode: AssalDataSourceMode;
  demoCatalog: DemoCatalog;
  productionGateway?: ProductionSelectGateway;
}): AssalRepository {
  if (options.mode === "demo") return new DemoRepository(options.demoCatalog);
  if (!options.productionGateway) throw new Error("Production repository requires an explicit Supabase gateway configuration.");
  return new ProductionRepository(options.productionGateway);
}

File: packages/data_dart/lib/repository_factory.dart
import 'assal_repository.dart';
import 'demo_repository.dart';
import 'production_repository.dart';

class AssalRepositoryFactory {
  const AssalRepositoryFactory();

  AssalRepository create({
    required AssalDataSourceMode mode,
    required DemoCatalogLoader demoLoader,
    ProductionQueryGateway? productionGateway,
  }) {
    switch (mode) {
      case AssalDataSourceMode.demo:
        return DemoRepository(loader: demoLoader);
      case AssalDataSourceMode.production:
        final gateway = productionGateway;
        if (gateway == null) throw const ProductionRepositoryNotConfigured();
        return ProductionRepository(gateway: gateway);
    }
  }
}
